serving/src/main/java/feast/auth/credentials/JwtAuthCredentials.java

```diff
@@ -0,0 +1,120 @@
+
+import static io.grpc.Metadata.ASCII_STRING_MARSHALLER;
+
+import io.grpc.CallCredentials;
+import io.grpc.Metadata;
+import io.grpc.Status;
+import java.time.Instant;
+import java.util.Map;
+import java.util.concurrent.Executor;
+import javax.security.sasl.AuthenticationException;
+import okhttp3.FormBody;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.RequestBody;
+import okhttp3.Response;
+import org.springframework.boot.configurationprocessor.json.JSONObject;
+import org.springframework.security.oauth2.jwt.NimbusJwtDecoder;
+
+/*
+ * Oauth Credentials Implementation for serving.
+ * Used by CoreSpecService to connect to core.
+ */
+public class JwtAuthCredentials extends CallCredentials {
+
+  private static final String JWK_ENDPOINT_URI = "jwkEndpointURI";
+  static final String APPLICATION_JSON = "application/json";
+  static final String CONTENT_TYPE = "content-type";
+  static final String BEARER_TYPE = "Bearer";
+  static final String GRANT_TYPE = "grant_type";
+  static final String CLIENT_ID = "client_id";
+  static final String CLIENT_SECRET = "client_secret";
+  static final String AUDIENCE = "audience";
+  static final String OAUTH_URL = "oauth_url";
+  static final Metadata.Key<String> AUTHORIZATION_METADATA_KEY =
+      Metadata.Key.of("Authorization", ASCII_STRING_MARSHALLER);
+
+  private OkHttpClient httpClient;
+  private Request request;
+  private String accessToken;
+  private Instant tokenExpiryTime;
+  private NimbusJwtDecoder jwtDecoder;
+
+  public JwtAuthCredentials(CoreAuthenticationProperties authProperties) {
+    this.httpClient = new OkHttpClient();
+    Map<String, String> options = authProperties.getOptions();
+    if (!(options.containsKey(GRANT_TYPE)
+        && options.containsKey(CLIENT_ID)
+        && options.containsKey(AUDIENCE)
+        && options.containsKey(CLIENT_SECRET)
+        && options.containsKey(OAUTH_URL)
+        && options.containsKey(JWK_ENDPOINT_URI))) {
+      throw new AssertionError(
+          "please configure the properties:"
+              + " grant_type, client_id, client_secret, audience, oauth_url, jwkEndpointURI");
+    }
+    RequestBody requestBody =
+        new FormBody.Builder()
+            .add(GRANT_TYPE, options.get(GRANT_TYPE))
+            .add(CLIENT_ID, options.get(CLIENT_ID))
+            .add(CLIENT_SECRET, options.get(CLIENT_SECRET))
+            .add(AUDIENCE, options.get(AUDIENCE))
+            .build();
+    this.request =
+        new Request.Builder()
+            .url(options.get(OAUTH_URL))
+            .addHeader(CONTENT_TYPE, APPLICATION_JSON)
+            .post(requestBody)
+            .build();
+    this.jwtDecoder = NimbusJwtDecoder.withJwkSetUri(options.get(JWK_ENDPOINT_URI)).build();
+  }
+
+  @Override
+  public void thisUsesUnstableApi() {
+    // TODO Auto-generated method stub
+
+  }
+
+  @Override
+  public void applyRequestMetadata(
+      RequestInfo requestInfo, Executor appExecutor, MetadataApplier applier) {
```

Comment:
I think it's still hard for me to validate whatever we are doing here without reference implementations or services. @jmelinav has shared https://auth0.com/docs/api-auth/tutorials/client-credentials, but ideally we would have more than one example to know whether what we are building is generalized. I would hate to have breaking changes in the future if we didn't get this right, and yes I know this functionality has already shipped in the Python SDK.

Have we looked at whether Keycloak supports this? Perhaps we can build an integration test around it.

Reply:
Yes Keycloak supports client credentials flow. It's an RFC defined in Oauth2 framework. 

- https://tools.ietf.org/html/rfc6749#section-4.4 
- https://www.keycloak.org/docs/latest/server_development/index.html#authenticate-with-a-service-account

Reply:
Hey @jmelinav. I think Keycloak would work, but it might be overkill. What if we tried to use Spring Security for this? https://docs.spring.io/spring-security-oauth2-boot/docs/current/reference/html5/

Reply:
That one is deprecated but I think spring security would be best https://github.com/spring-projects/spring-security/wiki/OAuth-2.0-Migration-Guide